src/v/datalake/translation_task.cc

```diff
@@ -0,0 +1,170 @@
+ */
+
+#include "datalake/translation_task.h"
+
+#include "datalake/logger.h"
+#include "datalake/record_multiplexer.h"
+
+#include <seastar/core/loop.hh>
+#include <seastar/core/seastar.hh>
+namespace datalake {
+namespace {
+remote_path calculate_remote_path(
+  const local_path& local_file_path, const remote_path& remote_path_prefix) {
+    auto f_name = local_file_path().filename();
+    return remote_path{remote_path_prefix() / f_name};
+}
+} // namespace
+translation_task::translation_task(cloud_data_io& cloud_io)
+  : _cloud_io(&cloud_io) {}
+
+ss::future<
+  checked<coordinator::translated_offset_range, translation_task::errc>>
+translation_task::translate(
+
+  std::unique_ptr<data_writer_factory> writer_factory,
+  model::record_batch_reader reader,
+  const remote_path& remote_path_prefix,
+  retry_chain_node& rcn,
+  lazy_abort_source lazy_as) {
+    record_multiplexer mux(std::move(writer_factory));
+    // Write local files
+    auto mux_result = co_await reader.consume(
+      std::move(mux), _read_timeout + model::timeout_clock::now());
+
+    if (mux_result.has_error()) {
+        vlog(
+          datalake_log.warn,
+          "Error writing data files - {}",
+          mux_result.error());
+        co_return errc::file_io_error;
+    }
+    auto write_result = std::move(mux_result.value());
+    vlog(
+      datalake_log.trace,
+      "translation result base offset: {}, last offset: {}, data files: {}",
+      write_result.start_offset,
+      write_result.last_offset,
+      fmt::join(write_result.data_files, ", "));
+
+    coordinator::translated_offset_range ret{
+      .start_offset = mux_result.value().start_offset,
+      .last_offset = mux_result.value().last_offset,
+    };
+    ret.files.reserve(write_result.data_files.size());
+    std::optional<errc> upload_error;
+    // TODO: parallelize uploads
+    for (auto& lf_meta : write_result.data_files) {
+        auto r = co_await execute_single_upload(
+          lf_meta, remote_path_prefix, rcn, lazy_as);
+        if (r.has_error()) {
+            vlog(
+              datalake_log.warn,
+              "error uploading file {} to object store - {}",
+              lf_meta,
+              r.error());
+            upload_error = r.error();
+            /**
+             * For now we value simplicity, therefore in case of cloud error we
+             * invalidate the whole translation i.e. we are going to cleanup all
+             * the local data files and remote files that were already
+             * successfully uploaded. Coordinator will simply retry translating
+             * the same range
+             */
+            break;
+        }
+        ret.files.push_back(coordinator::data_file{
+          .remote_path = r.value()().string(),
+          .row_count = lf_meta.row_count,
+          .file_size_bytes = lf_meta.size_bytes,
+          .hour = lf_meta.hour,
+        });
+    }
+
+    auto delete_result = co_await delete_local_data_files(
```

Comment:
should this be done in all cases (even if the lines above throw)? (wrap it in a as_future() or some such)

(or factor into a method called cleanup(), that the translator can call in all cases, after units are released, either is fine by me)

Reply:
it aculally is called in all the cases as the lines above do not return from the function.

Reply:
I thought upload_data_file() can potentially throw in cloud_io::upload_stream()?

Also the upload/cleanup could be done outside of the mutex, which can be plumbed into the task I guess.